vdr/doc/merge.go

```diff
@@ -0,0 +1,209 @@
+/*
+ * Nuts node
+ * Copyright (C) 2021 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package doc
+
+import (
+	"errors"
+	"sort"
+	"strings"
+
+	ssi "github.com/nuts-foundation/go-did"
+	"github.com/nuts-foundation/go-did/did"
+)
+
+// ErrDiffID is returned when a merge is done on documents that do not share the same ID
+var ErrDiffID = errors.New("no matching IDs")
+
+// MergeConflicted merges two DID Documents that share the same ID
+func MergeConflicted(docA did.Document, docB did.Document) (*did.Document, error) {
+	result := &did.Document{}
+	docs := []did.Document{docA, docB}
+
+	if !docA.ID.Equals(docB.ID) {
+		return nil, ErrDiffID
+	}
+
+	mergeBasics(docs, result)
```

Comment:
I'm thinking, everything (except `ID`) are basically sets that are merged into 1. Can't we simply unmarshal into a `map[string]interface{}` and merge those 2 maps? (you might still want/have to sort afterwards though).

Reply:
I was already looking for a way to reduce the similar code. This might be doable.

Reply:
This does more harm than good. Each method can contain an array with a map[string]interface{} or a reference which is just a string.
Unmarshalling in `map[string]interface{}` would also result in several casts on `interface{}` and we should also be aware for a combination of usages and different usages per doc. The resulting codes has less duplication but is way harder to read and thus to maintain.